dimod/include/dimod/iterators.h

```diff
@@ -37,14 +37,47 @@ class NeighborhoodIterator {
     using difference_type = std::ptrdiff_t;
     using iterator_category = std::random_access_iterator_tag;
 
+    // empty constructor needed for Cython
     NeighborhoodIterator()
             : neighborhood_ptr_(nullptr), i_(0), pair_ptr_(nullptr) {}
 
+    // for the copy constructor, we want to copy the pointer to the
+    // neighborhood, but not to the pair_ptr_, since that's managed by each
+    // iterator separately
+    NeighborhoodIterator(const NeighborhoodIterator<Bias, Index>& other)
+            : neighborhood_ptr_(other.neighborhood_ptr_),
+              i_(other.i_),
+              pair_ptr_(nullptr) {
+        update_pair();
+    }
+
+    // for the move constructor, we want to move everything
+    NeighborhoodIterator(NeighborhoodIterator<Bias, Index>&& other) noexcept
+            : neighborhood_ptr_(nullptr), i_(0), pair_ptr_(nullptr) {
+        swap(other);
+    }
+
     NeighborhoodIterator(neighborhood_type* neighborhood, index_type i)
             : neighborhood_ptr_(neighborhood), i_(i), pair_ptr_(nullptr) {
         update_pair();
     }
 
+    ~NeighborhoodIterator() { delete pair_ptr_; }
+
+    NeighborhoodIterator<Bias, Index>& operator=(
+            NeighborhoodIterator<Bias, Index> const& other) {
+        if (this == &other) return *this;
+        neighborhood_ptr_ = other.neighborhood_ptr_;
+        i_ = other.i_;
+        update_pair();  // handles pair_ptr_
+    }
+
+    NeighborhoodIterator<Bias, Index>& operator=(
```

Comment:
Dont you want to have the same check if (this == &other) like the other one ?

Reply:
Is swapping okay over here ?  or we need to use std::move ? and just before that delete the elements of the one being assigned to ?

Swapping makes sense for the move constructor though.

Reply:
I think so? https://stackoverflow.com/questions/276173/what-are-your-favorite-c-coding-style-idioms/2034447#2034447

Reply:
I think swap will swap the data, but I think we want the rhs side objects data to be invalidated. For cases such as these. 

 vector<int> func() {
    ....
    return vec;
}

x = func();

This is where we want the move constructor to work instead of the copy constructor, thus x will get all the data of the return value, but the return value should be invalidated and not take on the value of x, Thus what was inside x is not freed and that may cause memory leakage.

Reply:
I think it's ok, I see the same pattern in https://en.cppreference.com/w/cpp/language/rule_of_three

Reply:
Okay got it, since you have an explicit destructor, the temporary object will be destroyed and there wont be any memory leak.